gnomad_qc/v4/sample_qc/generate_qc_mt.py

```diff
@@ -0,0 +1,293 @@
+
+    samples_in_both = v4_mt.cols().semi_join(v3_mt.cols())
+    n_samples_in_both = samples_in_both.count()
+    if n_samples_in_both > 0:
+        logger.info("The following samples are found in both MatrixTables:")
+        samples_in_both.show(n_samples_in_both)
+        raise ValueError("Overlapping sample names were found in the MatrixTables!")
+
+    logger.info(
+        "Performing a union of the v3.1 and v4 predetermined QC site MatrixTable columns..."
+    )
+    mt = v3_mt.union_cols(v4_mt)
+
+    logger.info("Annotating with adj and filtering to pre LD-pruned QC sites...")
+    mt = annotate_adj(mt)
+
+    cp_path = get_checkpoint_path("combined_pre_ld_prune_qc_mt", mt=True)
+    # TODO: As I have it now, v3 and v4 are used to determine AF and callrate, is that OK, or should they be considered
+    #  independently and then shared passing variants merged?
+    mt = get_qc_mt(
+        mt,
+        bi_allelic_only=bi_allelic_only,
+        min_af=min_af,
+        min_callrate=min_callrate,
+        min_inbreeding_coeff_threshold=min_inbreeding_coeff_threshold,
+        min_hardy_weinberg_threshold=None,  # Already filtered from the initial set of QC variants
+        apply_hard_filters=False,
+        ld_r2=ld_r2,
+        checkpoint_path=cp_path,
+        filter_lcr=False,  # Already filtered from the initial set of QC variants
+        filter_decoy=False,  # Doesn't exist for hg38
+        filter_segdup=False,  # Already filtered from the initial set of QC variants
+    )
+    logger.info(
+        "Number of pre LD-pruned QC sites in gnomAD v3 + v4: %d...",
+        hl.read_matrix_table(cp_path).count_rows(),
+    )
+
+    return mt
+
+
+def main(args):
+    hl.init(
+        log="/generate_qc_mt.log",
+        default_reference="GRCh38",
+        tmp_dir="gs://gnomad-tmp-4day",
+    )
+    # NOTE: remove this flag when the new shuffle method is the default
+    hl._set_flags(use_new_shuffle="1")
+
+    overwrite = args.overwrite
+    ld_r2 = args.ld_r2
+    test = args.test
+    n_partitions = args.n_partitions
+
+    try:
+        if args.create_v3_filtered_dense_mt:
+            # Note: This command removes hard filtered samples
+            mt = get_gnomad_v3_mt(key_by_locus_and_alleles=True)
+            mt = create_filtered_dense_mt(mt, test, split=True)
+            mt = mt.checkpoint(
+                get_dense_predetermined_qc(version="3.1", test=test).path,
+                overwrite=overwrite,
+            )
+            logger.info(
+                "Number of predetermined QC variants found in the gnomAD v3 MatrixTable: %d...",
+                mt.count_rows(),
+            )
+
+        if args.create_v4_filtered_dense_mt:
+            # Note: This subset dense MatrixTable was created before the final hard filtering was determined
+            # Hard filtering is performed in `generate_qc_mt` before applying variant filters
+            vds = get_gnomad_v4_vds(split=True, remove_hard_filtered_samples=False)
```

Comment:
Yeah, I already created the dense MT with all samples since we didn't have the final hard filter and I don't want to spend the money to recreate it when I can just remove them before the final QC MT write.